plugins/boltz/swapmachine.go

```diff
@@ -0,0 +1,214 @@
+			Message:    fmt.Sprintf("Swap %d failed", in.GetJobID()),
+			IsError:    true,
+			IsFinished: true,
+		})
+	}
+	return FsmOut{}
+}
+
+func (s *SwapMachine) FsmSwapSuccess(in FsmIn) FsmOut {
+	if in.MsgCallback != nil {
+		in.MsgCallback(entities.PluginMessage{
+			JobID:      int32(in.GetJobID()),
+			Message:    fmt.Sprintf("Swap %d succeeded", in.GetJobID()),
+			IsError:    false,
+			IsFinished: true,
+		})
+	}
+	return FsmOut{}
+}
+
+func (s *SwapMachine) RedeemedCallback(data FsmIn, success bool) {
+	sd := data.GetSwapData()
+	if sd.State == RedeemingLockedFunds {
+		// even when this succeeds state is still screwed up
+		if success {
+			s.Eval(data, SwapFailed)
+		}
+	} else if sd.State == ClaimReverseFunds {
+		if success {
+			s.Eval(data, SwapSuccess)
+		} else {
+			s.Eval(data, SwapFailed)
+		}
+	}
+}
+
+// FsmWrap will just wrap a normal state machine function and give it the ability to transition states based on return values
+func FsmWrap[I FsmInGetter, O FsmOutGetter](f func(data I) O, b *Plugin) func(data I) O {
+	return func(in I) O {
+
+		realIn := in.Get()
+		out := f(in)
+		realOut := out.Get()
+
+		if realOut.Error != nil {
+			realOut.NextState = SwapFailed
+			realIn.MsgCallback(entities.PluginMessage{
+				JobID:      int32(realIn.GetJobID()),
+				Message:    fmt.Sprintf("Error %v happened", realOut.Error),
+				IsError:    true,
+				IsFinished: true,
+			})
+			return out
+		}
+
+		if realIn.SwapData.State.isFinal() {
+			return out
+		}
+		if realOut.NextState != None {
+			err := b.changeState(realIn, realOut.NextState)
+			realIn.MsgCallback(entities.PluginMessage{
+				JobID:      int32(realIn.GetJobID()),
+				Message:    fmt.Sprintf("Transitioning to state %v", realOut.NextState),
+				IsError:    false,
+				IsFinished: false,
+			})
+			if err != nil {
+				realOut.NextState = SwapFailed
+			}
+		}
+
+		return out
+	}
+}
+
+// State
+type State int
+
+const (
+	None State = iota
+
+	InitialForward
+	InitialReverse
+
+	SwapFailed
+	SwapSuccess
+
+	OnChainFundsSent
+	RedeemLockedFunds
+	RedeemingLockedFunds
+	VerifyFundsReceived
+
+	ReverseSwapCreated
+	ClaimReverseFunds
+)
+
+func (s State) String() string {
+	return []string{"None", "InitialForward", "InitialReverse", "SwapFaied", "SwapSuccess", "OnChainFundsSent",
+		"RedeemLockedFunds", "RedeemingLockedFunds", "VerifyFundsReceived", "ReverseSwapCreated", "ClaimReverseFunds"}[s]
+}
+
+func (s *State) isFinal() bool {
+	return *s == SwapFailed || *s == SwapSuccess
+}
+
+// Swapmachine is a finite state machine used for swaps.
+type SwapMachine struct {
+	Machine     *Fsm[FsmIn, FsmOut, State]
+	BoltzPlugin *Plugin
+}
+
+func NewSwapMachine(plugin *Plugin) *SwapMachine {
+	s := &SwapMachine{
+		Machine: &Fsm[FsmIn, FsmOut, State]{States: make(map[State]func(data FsmIn) FsmOut)},
+		// TODO: instead of BoltzPlugin this should be a bit more granular
+		BoltzPlugin: plugin,
+	}
+
+	s.Machine.States[SwapFailed] = FsmWrap(s.FsmSwapFailed, plugin)
```

Comment:
do mapping in one op, at line 186

Reply:
Thought about that, but can't because function depends on `s` instance.